### DISCUSSION TITLE
Feature Request: TOPN query

### DISCUSSION THREAD

**Discussion by kkalyan** [duckdb/duckdb#11511] [Ideas]
Why do you want this feature?
One of the common queries for analytics is get the top values for a column.
Adding percentage with a window function makes the output ready for sharing. 
Here is an example: 
```
SELECT
    page,
    count(1) as count,
    round(count(1)*100.0/sum(count(1)) over(),1) as percentage
FROM
    impressions
GROUP BY 1
ORDER BY 2 DESC
```

output
```
┌─────────┬──────────┬────────────┐
│ page.   │  count   │ percentage │
│ varchar │  int64   │   double   │
├─────────┼──────────┼────────────┤
│ page1   │   158073 │       73.0 │
│ page2+  │    58363 │       27.0 │
└─────────┴──────────┴────────────┘

```

this can be represented with simpler query saves 80% of key strokes. 
```
TOPN page FROM impressions [LIMIT N]
````

some analytics systems support this function
Druid - https://druid.apache.org/docs/latest/querying/topnquery
Microsoft DAX - https://learn.microsoft.com/en-us/dax/topn-function-dax

Another variant of this query is aggregate by a column value rather than counting it. 
For example if a table `impressions` contains a column `clicks` and the user wants to know click percentage by a feature `page`

```
SELECT
         page,
         sum(clicks) as clicks, 
         round(sum(clicks)*100.0/sum(sum(clicks)) over(),2) as clicks_percentage
FROM
           impressions
GROUP BY 1
ORDER BY 2 DESC
LIMIT N
```

with output 
```
┌─────────┬───────────┬───────────────────┐
│  page   │  clicks   │ clicks_percentage │
│ varchar │  double   │      double       │
├─────────┼───────────┼───────────────────┤
│ page1   │ 6691978   │             97.46 │
│ page2+  │  174312   │              2.54 │
└─────────┴───────────┴───────────────────┘
```

which can be represented by 
```
TOPN page [BY clicks] FROM impressions [LIMIT N]
````

Not sure if this can be achieved by duckdb macros, but this goes well with the theme of DuckDb simplifying the SQL further.

**Comment by Tishj** [comment]
Sounds good, we already have a TopN operator, inserted by an optimizer rule
This would just require parser, transformer and binder support

**Comment by soerenwolfers** [comment]
Personally, I'm not convinced that this has a good complexity-value proposition.

**Reply by kkalyan** [reply]
I understand the confusion the second query caused. The first query is probably most common type of  analytic workloads and complexity is worth the usability it brings. While calculating percentages may be trivial, I’ve seen many folks using excel/pandas to compute percentages as they are comfortable using window functions. I hope I made the argument clear for friendlier SQL.

**Reply by soerenwolfers** [reply]
Ah, sorry, I didn't read your examples right the first time.
